### PR TITLE
pfSense-pkg-bind9 Add support for response policy zones

### PR DESCRIPTION
**Pull request by miken32**
More info on RPZs:
https://dnsrpz.info/
http://www.zytrax.com/books/dns/ch7/rpz.html
https://www.isc.org/rpz/

### CODE REVIEW

**Comment by miken32** [comment]
Note I've not tested this with a slave RPZ; I don't have the setup to do so. Works great with a master zone.

**Comment by jim-p** [comment]
In an effort to improve the tracking of changes and bug/feature requests, we have decided to require an entry on the [pfSense Redmine issue tracker](https://redmine.pfsense.org) associated with every pull request, and likewise the Redmine entry should also have a link back to the pull request.

If you could, please associate this PR to a Redmine issue either by locating an existing issue at https://redmine.pfsense.org or by creating a new issue. Add a link to the Redmine issue that points to this PR, and also add a link on the PR to the Redmine issue.

For more information, see https://doc.pfsense.org/index.php/Submitting_a_Pull_Request_via_Github

Thanks!

**Comment by miken32** [comment]
[Redmine feature request](8188 https://redmine.pfsense.org/issues/8188)

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind_zones.xml [hunk 1]
```diff
@@ -228,8 +234,8 @@
 			<fieldname>temp03</fieldname>
 		</field>
 		<field>
-			<fielddescr>TLL</fielddescr>
-			<fieldname>tll</fieldname>
+			<fielddescr>TTL</fielddescr>
+			<fieldname>ttl</fieldname>
```
Comment (jim-p) [inline dns/pfSense-pkg-bind9/files/usr/local/pkg/bind_zones.xml]: Changing the field should also be accompanied by some code to carry over the user's old setting for the field. Otherwise, the user's existing value will be lost.

The TLL/TTL correction doesn't appear to be related to the RPZ changes in this PR, is there some reason they need to happen together? If not, the TLL/TTL change should be its own PR so the two sets of changes are distinct.

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc [hunk 2]
```diff
@@ -43,9 +43,14 @@ function bind_zone_validate($post, &$input_errors) {
 		$_POST['mail'] = preg_replace("/@/", ".", $post['mail']);
 	}
 
+	if (isset($_POST['rpz']) && $_POST['rpz'] == "on") {
+		if ($_POST['type'] !== "master" && $_POST['type'] !== "slave") {
```
Comment (jim-p) [inline dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc]: In this line and the line above, please put the individual tests with multiple components inside parenthesis, for example:
`if (isset($_POST['rpz']) && ($_POST['rpz'] == "on")) {`
and
`if (($_POST['type'] !== "master") && ($_POST['type'] !== "slave")) {`

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc [hunk 3]
```diff
@@ -140,6 +145,17 @@ function bind_sync() {
...
 
 	$bind = $config['installedpackages']['bind']['config'][0];
+	if (is_array($config['installedpackages']['bindzone'])) {
+		$bindzone = $config['installedpackages']['bindzone']['config'];
+	} else {
+		$bindzone = array();
+	}
+	$rpz_zones = array();
+	foreach ($bindzone as $zone) {
+		if (isset($zone['rpz']) && $zone['rpz'] == "on") {
```
Comment (jim-p) [inline dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc]: See previous notes about parenthesis in `if` statement tests.

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc [hunk 4]
```diff
@@ -163,6 +179,13 @@ options {
 	max-cache-size {$ram_limit};
 
 EOD;
+	if (count($rpz_zones) > 0) {
+		$bind_conf .= "\tresponse-policy {\n";
+		foreach ($rpz_zones as $rpz) {
+			$bind_conf .= "\t\tzone \"$rpz\";\n";
```
Comment (jim-p) [inline dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc]: Variables embedded in strings must be enclosed in braces, for example:
`$bind_conf .= "\t\tzone \"{$rpz}\";\n";`

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc [hunk 5]
```diff
@@ -356,6 +373,7 @@ EOD;
 			}
 			$zonereverso = $zone['reverso'];
 			$zonereversv6o = $zone['reversv6o'];
+			$zonerpz = (isset($zone["rpz"]) && $zone["rpz"] === "on");
```
Comment (jim-p) [inline dns/pfSense-pkg-bind9/files/usr/local/pkg/bind.inc]: See previous notes about parenthesis in `if` statement tests.

**Comment by miken32** [comment]
Review comments addressed. I reverted the change to the config field name, and made the coding standards changes to my code (as well as the existing, so it should be consistent now.)

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind_zones.xml [hunk 6]
```diff
@@ -122,7 +122,7 @@
 			<type>select</type>
 			<options>
 				<option><name>Master</name><value>master</value><enablefields>description</enablefields></option>
-				<option><name>Slave</name><value>slave</value><enablefields>ttl</enablefields></option>
+				<option><name>Slave</name><value>slave</value><enablefields>tll</enablefields></option>
```
Comment (jim-p) [inline dns/pfSense-pkg-bind9/files/usr/local/pkg/bind_zones.xml]: Just this one part needs set back to TTL and then everything should be OK

**Comment by miken32** [comment]
Actually the field is still called `tll`, I just fixed that line to match the existing typo.

File: dns/pfSense-pkg-bind9/files/usr/local/pkg/bind_zones.xml (hunk 6 above)
Comment (miken32) [reply dns/pfSense-pkg-bind9/files/usr/local/pkg/bind_zones.xml]: Actually the field is still called tll, I just fixed that line to match the existing typo.

**Comment by jim-p** [comment]
OK, I must have misread it. Looks good now then, thanks!
